### ISSUE TITLE
Using multiple sections kills performance

### ISSUE THREAD

**Issue by mpcjanssen** [bytesparadise/libasciidoc#121]
Using even a small number of sections seems to encounter a bad case in the parser as performance reduces drastically.

For example for this simple file

````
=== 1
 dasda

=== 2
 dasda

=== 3
 dasda

=== 4
 dasda

=== 5
 dasda

=== 6
 dasda

=== 7
 dasda

=== 8
 dasda

=== 9
 dasda

=== 10
````

takes much longer to parse then using the ruby version of asciidoctor

````
$ time libasciidoc -s test.adoc > /dev/null

real	0m3.806s
user	0m4.255s
sys	0m0.082s

$ time asciidoctor test.adoc > /dev/null

real	0m0.620s
user	0m0.264s
sys	0m0.098s
````

**Comment by mpcjanssen** [comment]
Strangely enough I don't get any output when using the CLI command. However using the renderer in my own program it is much faster and gives the correct output.

**Comment by xcoulon** [comment]
@mpcjanssen Great news! I also changed a few things in the CLI to align with asciidoc/asciidoctor, so you may need to use the `-o -` option to see the generated HTML in `stdout`. Otherwise, you should have a `.html` file named after the `.adoc` source file that you passed in the command. Does this help ? 

@mpcjanssen also, please feel free to report any other issue if you have any! Users' feedback is very helpful!

**Comment by mpcjanssen** [comment]
@xcoulon file was indeed created with a `html` extension. Thanks for the quick fix.

**Comment by mpcjanssen** [comment]
@xcoulon for a bit of context: I am displaying nested XML results on a web page. Instead of manually parsing all the xml and creating the HTML from scratch, I walk the XML and change it to asciidoc format. I then use this lib to generate the HTML.

So because parsing and rendering is done on every page call, it really needs to be fast.

**Comment by xcoulon** [comment]
re-opening to work on the grammar. The section part can be refactored to significantly improve the performances

**Comment by xcoulon** [comment]
TL;DR: In summary, parsing a document of 10 sections now takes `0.070s` vs `5.828s` before https://github.com/bytesparadise/libasciidoc/commit/1167a2033e4751a6b67eb41feb1a16120ab2ca30, i.e, more than `80x` faster.

Adding a few results comparison about https://github.com/bytesparadise/libasciidoc/commit/1167a2033e4751a6b67eb41feb1a16120ab2ca30, using https://github.com/bytesparadise/libasciidoc/blob/master/pkg/parser/bench_test.go backported to `master` branch

Bench running on `master`, prior to grammar changes:

````
• [MEASUREMENT]
parser benchmark
  bench parser on 10 lines

  Ran 10 samples:
  expr count - Smallest: 41424582.000,  Average: 41424582.000 ± 0.000, Largest: 41424582.000
  runtime - Fastest Time: 5.612s,  Average Time: 5.828s ± 0.422s,  Slowest Time: 7.077s
------------------------------
• [MEASUREMENT]
parser benchmark
  bench parser on 2 lines

  Ran 10 samples:
  expr count - Smallest: 16874.000,  Average: 16874.000 ± 0.000,  Largest: 16874.000
  runtime - Fastest Time: 0.002s,  Average Time: 0.002s ± 0.000s,  Slowest Time: 0.003s
------------------------------
• [MEASUREMENT]
parser benchmark
  bench parser on 1 line

  Ran 1 samples:
  expr count - Smallest: 5232.000,  Average: 5232.000 ± 0.000,  Largest: 5232.000
  runtime - Slowest Time: 0.001s,  Average Time: 0.001s ± 0.000s,  Fastest Time: 0.001s
````


Same bench, running with the grammar changes:

````
• [MEASUREMENT]
parser benchmark bench parser on 1 line
/Users/xcoulon/code/go/src/github.com/bytesparadise/libasciidoc/pkg/parser/bench_test.go:72

  1 samples:
    expr count - Smallest: 1374.000,  Average: 1374.000 ± 0.000,  Largest: 1374.000
    runtime - Fastest Time: 0.004s, Average Time: 0.004s ± 0.000s, Slowest Time: 0.004s
------------------------------
• [MEASUREMENT]
parser benchmark bench parser on 10 lines
/Users/xcoulon/code/go/src/github.com/bytesparadise/libasciidoc/pkg/parser/bench_test.go:14

  10 samples:
    expr count - Smallest: 19228.000,  Average: 19228.000 ± 0.000,  Largest: 19228.000
    runtime - Fastest Time: 0.066s, Average Time: 0.070s ± 0.003s, Slowest Time: 0.076s
------------------------------
• [MEASUREMENT]
parser benchmark bench parser on 2 lines
/Users/xcoulon/code/go/src/github.com/bytesparadise/libasciidoc/pkg/parser/bench_test.go:55

  10 samples:
    expr count - Smallest: 3416.000,  Average: 3416.000 ± 0.000,  Largest: 3416.000
    runtime - Fastest Time: 0.009s, Average Time: 0.011s ± 0.002s, Slowest Time: 0.016s
````

**Comment by mpcjanssen** [comment]
Great work 👍

**Comment by xcoulon** [comment]
thanks, @mpcjanssen!
